KerasNLP : `Semantic similarity` tutorial

shivance opened the pull request:
Issue : [KerasNLP/#726](https://github.com/keras-team/keras-nlp/issues/726)
Kaggle Tutorial [Link](https://www.kaggle.com/code/shivanshuman/kerasnlp-semantic-similarity-with-bert/notebook)

This PR ports Keras-io's existing [Semantic Similarity with BERT](https://keras.io/examples/nlp/semantic_similarity_with_bert/) to KerasNLP. Now the tutorial is pure Keras + tensorflow. It doesn't depend on Tensorflow at all.

shivance commented:
cc: @mattdangerw

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+"""
+Semantic Similarity with KerasNLP
```
This should be a `title:` field

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+"""
+
+
+
```
Please remove extra empty lines

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+
+
+"""
+Introduction
```
Section titles should start with `## `

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+"""
+Introduction
+
+Semantic Similarity is the task of determining how similar two sentences are, in terms of what they mean. This example demonstrates the use of SNLI (Stanford Natural Language Inference) Corpus to predict sentence semantic similarity with Transformers. We will fine-tune a BERT model that takes two sentences as inputs and that outputs a similarity score for these two sentences. This notebook reprod ...
```
Please keep text lines under 80 chars

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+import keras_nlp
+
+# Import classes corresponding to BERT model from KerasNLP
+from keras_nlp.models.bert.bert_preprocessor import BertPreprocessor
```
Please import from `bert` instead

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+labels = ["contradiction", "entailment", "neutral"]
+
+
+# In[8]:
```
Please remove notebook annotations like this one

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+KerasNLP offers pretrained models, we'll use the same here
+"""
+
+model = BertClassifier.from_preset(
```
Please run `black` on the file

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+# In just four epochs we achieved 71% of accuracy.
+
+"""
+### Evaluate model on the test set
```
Please add a section showing:

1. saving the model
2. reloading the model
3. performing inference with the model

fchollet commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,169 @@
+"""
+### Train the Model End to End
+ 
+KerasNLP offers pretrained models, we'll use the same here
```
We need more polished text commentary explaining what is going on, what the best practices are, what the reader should take away from it, etc.

fchollet reviewed (commented):
Thanks for the PR!

mattdangerw commented:
Thanks for the ping! Need to get better at checking keras.io PRs. Will review today!

mattdangerw commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,216 @@
+    epochs: The number of epochs to train the model
+    batch_size: The number of samples per gradient update
+"""
+model.fit(x=x_train, y=y_train, validation_data=(x_val, y_val), epochs=4, batch_size=64)
```
I wonder if it's worth a plug to show how easy it would be to switch to other models to boost performance. E.g. show setting up a run with `roberta_base_en` and talk about how much that would improve things.

mattdangerw reviewed (changes requested):
Looks good! But we need to make sure this stand on it's own as a separate example from https://keras.io/examples/nlp/semantic_similarity_with_bert/, left a few suggestion on that front.

shivance commented:
Hi @mattdangerw ! 
I tried to address all comments. Please review & let me know if anything else needs to be changed. Thanks.

shivance commented:
> Hi @mattdangerw ! I tried to address all comments. Please review & let me know if anything else needs to be changed. Thanks.
> 
> Corresponding Notebook can be found [here](https://www.kaggle.com/code/shivanshuman/kerasnlp-semantic-similarity-with-bert). Also you might find this [thread](https://twitter.com/anshum0n/status/1633548651787333633?s=20) interesting 😄 .

@mattdangerw  btw instead of just tf.data.Dataset, I used tensorflow_datasets library, totally eliminating need to download explicitly. Now the stack is pure tf :)

shivance commented on guides/keras_nlp/semantic-similarity-with-bert.py:
```diff
@@ -0,0 +1,216 @@
+"""
+
+model = BertClassifier.from_preset(
+    "bert_tiny_en_uncased", num_classes=3, preprocessor=None
```
We need to keep preprocessor=None here in the new setting with tfds.
See last cells of this [colab](https://colab.research.google.com/drive/1KCNxATaeNcmxk-XJcgYBAFb4gngk5_rf?usp=sharing)

shivance commented:
Hi @mattdangerw @fchollet  ! 
I tried to address all comments. Please review & let me know if anything else needs to be changed. Thanks.

Corresponding Notebook can be found [here](https://www.kaggle.com/code/shivanshuman/kerasnlp-semantic-similarity-with-bert). Also you might find this [thread](https://twitter.com/anshum0n/status/1633548651787333633?s=20) interesting 😄 .

shivance commented:
Hey @mattdangerw, The CI looks green. I think it's ready for another round of review.

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+
+"""
+
+snli_train = tfds.load("snli", split="train[:20%]")
```
very cool! I didn't know about this trick

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+"""
+
+
+def preprocess_dataset(dataset, preprocessor):
```
Seems a little inefficient to go to pandas and back, and I'm not sure we really need it. Are we doing anything here besides filtering some labels? https://www.tensorflow.org/api_docs/python/tf/data/Dataset#filter should do the trick

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+"""
+Let's take a look at how our dataset looks like
+"""
+snli_val_df = to_dataframe(snli_val)
```
You can probably do something like `snli_val.batch(2).take(1).get_single_element()` to preview a dataset (if we ditch the dataframe code here)

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+to use your own preprocessor. Here we'll set the `num_classes` as 3 for SNLI dataset
+"""
+
+bert_clf = keras_nlp.models.BertClassifier.from_preset(
```
I would avoid clf as an abbreviation here. Either lets use `classifier = ..` everywhere or `xx_classifier = ...` everywhere.

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+x_val, y_val = preprocess_dataset(snli_val, bert_preprocessor)
+
+"""
+### Train the Model End to End
```
End-to-end

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+bert_clf.summary()
+
+"""
+KerasNLP models compile automatically. Let's train the model we just instantiated, using 
```
KerasNLP models compile automatically. -> KerasNLP task models come with compilation defaults.

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+    # some sample have -1 as label entries. We'll simply drop those here
+    dataset.drop(dataset[dataset["label"] == -1].index, inplace=True)
+
+    x = preprocessor((dataset["hypothesis"], dataset["premise"]))
```
If you set up your dataset correctly, you should be able to just do
`dataset = dataset.map(preprocessor, num_parallel_calls=tf.data.AUTOTUNE)`

The results `(x, y)` tuple dataset can be passed directly to fit as a single entity.

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+
+bert_clf_1.compile(
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    optimizer=tf.keras.optimizers.Adam(1e-4),
```
Maybe show off AdamW here? That should boost performance somewhat.

Also you could try showing a learning rate with warmup and linear decay. I'm not sure that will show too much different with the "tiny" checkpoints, but with larger ones it should.

Here's a simple learning rate schedule I have used in the past.

```

class TriangularSchedule(keras.optimizers.schedules.LearningRateSchedule):
    """Linear ramp up for `warmup` steps, then linear decay to zero at `total` steps."""
    def __init__(self, rate, warmup, total):
        self.rate = tf.cast(rate, dtype="float32")
        self.warmup = tf.cast(warmup, dtype="float32")
        self.total = tf.cast(total, dtype="float32")

    def __call__(self, step):
        step = tf.cast(step, dtype="float32")
        multiplier = tf.cond(
            step < self.warmup,
            lambda: step / self.warmup,
            lambda: (self.total - step) / (self.total - self.warmup),
        )
        return tf.maximum(self.rate * multiplier, 0.0)
```

This might be useful to demo, as it's a very common learning rate schedule in NLP.

shivance replied:
Thanks for the input, with bert tiny x triangular rate scheduler, val and test accuracy hikes up to 85%

shivance replied:
Also AdamW on 1st epoch with a LR scheduler gives 65% of val accuracy, way lesser than adam, moving with Adam only here, what say @mattdangerw ?

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+
+Let's recompile our model with a different learning rate and see performance
+"""
+bert_clf_1 = keras_nlp.models.BertClassifier.from_preset(
```
Maybe just redefine `bert_classifier` here. Overall comment is to keep variable names readable!

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+
+# Save and Reload the model
+"""
+bert_clf.save("bert_classifier.pb")
```
Do we want to show this format for saving? Should we just show `bert_clf.save("bert_classifier")`?

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+
+Now that we have established a baseline, we'll attempt to get better results by trying different
+models. KerasNLP makes experimentation easy for us, with just few lines of code, we can train 
+a roberta on the same dataset.
```
train a roberta -> fine-tune a RoBERTa checkpoint

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+print(tf.math.argmax(predictions, axis=1).numpy())
+
+"""
+# Enhancing accuracy with RoBERTA
```
RoBERTa

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+x_test, y_test = preprocess_dataset(snli_test, roberta_preprocessor)
+x_val, y_val = preprocess_dataset(snli_val, roberta_preprocessor)
+
+# inittializing a roberta from preset
```
inittializing -> Initializing

(also captilize start of comments above too)

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+roberta_clf.evaluate(x=x_test, y=y_test)
+
+"""
+`robeta_base_en` is slightly bigger model than bert_tiny, it took almost 1.5 hrs to train on
```
roberta_base_en is a much bigger model right? We might want to call out exactly how much the parameter counts are changing, I think you are growing the model 30-fold!

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,274 @@
+validation and test split. 16 was the biggest batch size that we could fit on Kaggle P100 GPU
+with roberta as our model.
+
+The steps to perform inference with roberta remain same as our bert !
```
The steps to perform inference with the RoBERTa model remain same as with BERT!

mattdangerw reviewed (changes requested):
Nice! This is flowing better.

Comments below, but the main one is we don't seem to have much of a reason to reach for pandas here, and we might want to dig into more options for compilation (I think that would be of genuine interest to users).

Thanks very much excited to add this!

shivance commented:
Hi @mattdangerw , I tried eliminating usage of pandas entirely, and making entire workflow with tf.data only

https://www.kaggle.com/code/shivanshuman/kerasnlp-semantic-similarity-with-bert

but for some reason, it's showing nan loss everywhere. I can't seem to figure out why. Can you please take a quick look?

mattdangerw commented:
Left a comment https://github.com/keras-team/keras-nlp/issues/880#issuecomment-1477245047, hope that helps.

shivance commented:
Hey @mattdangerw , it's ready for review, Phewww 😌 
Sorry for delay !

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,292 @@
+bert_classifier.compile(
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    optimizer=tf.keras.optimizers.experimental.AdamW(
+        TriangularSchedule(0.001, 1000, 5000)
```
Maybe we should show how to get an accurate count of the train steps here. It's a bit tricky, but useful to show to users. Something like this (only first two lines are new)...

```python
# Get the total count of training batches.
# This requires walking the dataset to filter all -1 labels.
total_steps = sum(1 for _ in train_ds.as_numpy_iterator())
warmup_steps = int(total_steps * 0.2)

bert_classifier.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer=tf.keras.optimizers.experimental.AdamW(
        TriangularSchedule(0.001, warmup_steps, total_steps)
    ),
    metrics=["accuracy"],
)
```

shivance replied:
Hey @mattdangerw ! cool !
but please take a look at following:

![image](https://user-images.githubusercontent.com/51750587/227028150-25da5a5a-92e6-4c8e-964e-208eff221383.png)

shivance replied:
It saturates.

mattdangerw replied:
Ah, I might have made a mistake here, I think we want to do this rate across the total number of train steps for *all epochs*. If you update to this, how do things look?

```python
epochs = 3
# Get the total count of training batches.
# This requires walking the dataset to filter all -1 labels.
total_steps = sum(1 for _ in train_ds.as_numpy_iterator()) * epochs
warmup_steps = int(total_steps * 0.2)
```

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,292 @@
+predictions = bert_classifier.predict(sample)
+
+# Get the class predictions with maximum probabilities
+print(tf.math.argmax(predictions, axis=1).numpy())
```
Should we show softmax rather than argmax here? So we show that these are probabilities coming out?

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,292 @@
+"""
+
+predictions = roberta_classifier.predict(sample)
+print(tf.math.argmax(predictions, axis=1).numpy())
```
Same question here, softmax rather than argmax?

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,292 @@
+
+
+import numpy as np
+import pandas as pd
```
pandas is now unused i think

mattdangerw commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,292 @@
+"""
+
+
+import numpy as np
```
numpy also looks unused.

mattdangerw reviewed (approved):
This is great! Left a few more suggestions, but overall this has a great flow and will be super useful to readers I think.

Dropping an approval so @fchollet can merge whenever comments are addressed and things look good to him.

shivance commented:
Addressed the commits, @fchollet could you review and let me know if any comments?

shivance commented:
Hi @fchollet , I would request your review on this PR, Thanks.

fchollet commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,297 @@
+"""
+
+"""
+# Introduction
```
All sections should use `##` -- the main title will be autogenerated.

fchollet commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,297 @@
+Author: [Anshuman Mishra](https://github.com/shivance/)
+Date created: 2023/02/25
+Last modified: 2023/02/25
+Description: Use pretrained models from KerasNLP for Semantic Similarity Task
```
"for the Semantic Similarity task."

fchollet commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,297 @@
+learn how to use [KerasNLP](https://keras.io/keras_nlp/), an extension of the core Keras API, 
+for the same task. We'll also learn how KerasNLP reduces the boilerplate code and makes models easy to use.
+
+This guide is broken into following parts:
```
broken down

fchollet commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,297 @@
+
+Semantic Similarity is the task of determining how similar two sentences are, in terms of 
+what they mean. We already saw in [this](https://keras.io/examples/nlp/semantic_similarity_with_bert/) 
+example how to use SNLI (Stanford Natural Language Inference) Corpus to predict sentence 
```
"corpus (lowercase)

fchollet commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,297 @@
+Semantic Similarity is the task of determining how similar two sentences are, in terms of 
+what they mean. We already saw in [this](https://keras.io/examples/nlp/semantic_similarity_with_bert/) 
+example how to use SNLI (Stanford Natural Language Inference) Corpus to predict sentence 
+semantic similarity with HuggingFace Transformers library. In this tutorial we will 
```
"with the"

fchollet commented on examples/nlp/semantic_similarity_with_keras_nlp.py:
```diff
@@ -0,0 +1,297 @@
+)
+
+"""
+KerasNLP models automatically tokenizes the inputs as per the model used, but user can use custom preprocessing
```
"tokenize" (no s).

fchollet reviewed (commented):
The code looks good! There are various small issues with the text copy. Can you run each text paragraph into ChatGPT and ask it for typos fixes and copyedits?

shivance commented:
My git is messed up. Probably I need to uninstall and reinstall. Closing this PR from here, I have addressed the comments and opened a new PR #1299 

Sorry for the inconvenience.
